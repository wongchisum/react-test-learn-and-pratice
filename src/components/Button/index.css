/* 通用样式 */
.btn {
    border: 1px solid #d9d9d9;
    line-height: 1em;
    border-radius: 6px;
    background-color: #fff;
    cursor: pointer;
    text-align: center;
}

/* 布局样式 */
.inline-block {
    display: inline-block;
}

.block {
    display: block;
}

/* 配色样式 */
.primary {
    background-color: #09d;
    color: #fff;
}
.danger {
    background-color: #ff0000;
    color: #fff;
}
.info {
    background-color: green;
    color: #fff;
}

/* 尺寸样式 */

.sm {
    font-size: 12px;
    padding: 4px 8px;
}

.md {
    font-size: 14px;
    padding: 6px 10px;
}

.lg {
    font-size: 16px;
    padding: 8px 12px;
}