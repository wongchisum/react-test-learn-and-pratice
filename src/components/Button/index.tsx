// import "./index.css";
import type { ButtonProps } from "./types";
import { getClassFromSchema } from "./helper";
import { isType } from "../../utils/type";
import {
  ButtonLayout,
  ButtonSize,
  ButtonTheme,
  DefaultLayout,
  DefaultSize,
  DefaultTheme,
} from "./constant";
import { SyntheticEvent } from "react";
const _isFn = isType("function");

export const Button = (props: ButtonProps) => {
  const { theme, size, children, layout, onClick } = props;
  const sizeClass = getClassFromSchema(DefaultSize, ButtonSize, size);
  const themeClass = getClassFromSchema(DefaultTheme, ButtonTheme, theme);
  const layoutSize = getClassFromSchema(DefaultLayout, ButtonLayout, layout);

  const onClickWrap = (event: SyntheticEvent<HTMLElement>) => {
    event.preventDefault();
    _isFn(onClick) && (onClick as Function)(event);
  };
  return (
    <div
      className={`btn ${sizeClass} ${themeClass} ${layoutSize}`}
      onClick={onClickWrap}
    >
      {children}
    </div>
  );
};
