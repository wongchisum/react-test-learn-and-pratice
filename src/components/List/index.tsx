import type { ListProps } from "./types";
import { execRender } from "./helper";

export function List<T>(props: ListProps<T>) {
  const { data, rowRender } = props;
  if (Array.isArray(data)) {
    return (
      <div>
        {data.map((item: T, index: number) => {
          const element = execRender(rowRender, item, index);
          return element && <div key={index}>{element}</div>;
        })}
      </div>
    );
  }
  return null;
}
