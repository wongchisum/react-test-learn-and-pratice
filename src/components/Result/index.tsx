import type { ResultProps } from "./types";
import { getResult, isNull } from "./helper";

export function Result(props: ResultProps) {
  const { success, successContent, failContent } = props;
  const resultState = getResult(success);
  const successView = isNull(successContent)
    ? null
    : successContent || "Success!";
  const failView = isNull(failContent) ? null : failContent || "Fail!";
  return <div>{resultState ? successView : failView}</div>;
}
