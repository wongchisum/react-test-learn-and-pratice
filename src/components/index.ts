export { List } from "./List";
export { Result } from "./Result";
export { Button } from "./Button";
