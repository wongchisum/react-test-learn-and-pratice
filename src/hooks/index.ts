export { useCounter } from "./useCounter";
export { useBool } from "./useBool";
