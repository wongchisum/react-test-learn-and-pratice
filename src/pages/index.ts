export { default as ListPage } from "./ListPage";
export { default as ButtonPage } from "./ButtonPage";
