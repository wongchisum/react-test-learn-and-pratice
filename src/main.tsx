import React from "react";
import ReactDOM from "react-dom";
import { ButtonPage } from "./pages";

const App = () => {
  return (
    <div>
      <ButtonPage />
    </div>
  );
};

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById("root")
);
