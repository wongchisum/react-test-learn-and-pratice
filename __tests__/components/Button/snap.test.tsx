import renderer from "react-test-renderer";
import { Button } from "../../../src/components";

// 进行快照测试，会生成__snapshots__文件夹和.snap文件
// 描述测试用例渲染json结果
// 如果snap结果发生变动，可能是组件变动导致的bug，会导致报错
// 如果是组件的内部需要变动，可以使用npx jest -u进行快照的刷新

const Text = () => {
  return <div>Hello World</div>;
};

describe("测试Button组件props", () => {
  it("default", () => {
    const view = renderer.create(<Button></Button>).toJSON();
    expect(view).toMatchSnapshot();
  });

  it("With children props", () => {
    const textView = renderer.create(<Button>Text</Button>).toJSON();
    expect(textView).toMatchSnapshot();

    const elementView = renderer
      .create(
        <Button>
          <div className="element">Hello</div>
        </Button>
      )
      .toJSON();
    expect(elementView).toMatchSnapshot();

    const fcView = renderer
      .create(
        <Button>
          <Text />
        </Button>
      )
      .toJSON();
    expect(fcView).toMatchSnapshot();
  });

  it("With theme props", () => {
    const dangerView = renderer
      .create(<Button theme="danger">danger</Button>)
      .toJSON();
    expect(dangerView).toMatchSnapshot();

    const infoView = renderer
      .create(<Button theme="info">info</Button>)
      .toJSON();
    expect(infoView).toMatchSnapshot();

    const primaryView = renderer
      .create(<Button theme="primary">primary</Button>)
      .toJSON();
    expect(primaryView).toMatchSnapshot();
  });
  it("With layout props", () => {
    const inlineView = renderer
      .create(<Button layout="inline-block">inline-block</Button>)
      .toJSON();
    expect(inlineView).toMatchSnapshot();

    const blockView = renderer
      .create(<Button layout="block">block</Button>)
      .toJSON();
    expect(blockView).toMatchSnapshot();
  });
});
