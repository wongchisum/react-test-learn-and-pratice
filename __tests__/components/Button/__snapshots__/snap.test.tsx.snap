// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Button组件props With children props 1`] = `
<div
  className="btn md default inline-block"
  onClick={[Function]}
>
  Text
</div>
`;

exports[`测试Button组件props With children props 2`] = `
<div
  className="btn md default inline-block"
  onClick={[Function]}
>
  <div
    className="element"
  >
    Hello
  </div>
</div>
`;

exports[`测试Button组件props With children props 3`] = `
<div
  className="btn md default inline-block"
  onClick={[Function]}
>
  <div>
    Hello World
  </div>
</div>
`;

exports[`测试Button组件props With layout props 1`] = `
<div
  className="btn md default inline-block"
  onClick={[Function]}
>
  inline-block
</div>
`;

exports[`测试Button组件props With layout props 2`] = `
<div
  className="btn md default block"
  onClick={[Function]}
>
  block
</div>
`;

exports[`测试Button组件props With theme props 1`] = `
<div
  className="btn md danger inline-block"
  onClick={[Function]}
>
  danger
</div>
`;

exports[`测试Button组件props With theme props 2`] = `
<div
  className="btn md info inline-block"
  onClick={[Function]}
>
  info
</div>
`;

exports[`测试Button组件props With theme props 3`] = `
<div
  className="btn md primary inline-block"
  onClick={[Function]}
>
  primary
</div>
`;

exports[`测试Button组件props default 1`] = `
<div
  className="btn md default inline-block"
  onClick={[Function]}
/>
`;
